applications/minotari_merge_mining_proxy/src/proxy/mod.rs

```diff
@@ -0,0 +1,312 @@
+        "44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A";
+
+    async fn get_json_rpc(
+        method: MonerodMethod,
+        json_rpc_port: u16,
+        hyper_json_response: bool,
+        hash: Option<String>,
+        block_blob: Option<String>,
+    ) -> Result<String, Error> {
+        match method {
+            MonerodMethod::GetHeight => get_response(json_rpc_port, method).await,
+            MonerodMethod::GetBlockTemplate |
+            MonerodMethod::GetVersion |
+            MonerodMethod::SubmitBlock |
+            MonerodMethod::GetBlockHeaderByHash |
+            MonerodMethod::GetLastBlockHeader |
+            MonerodMethod::GetBlock => {
+                json_rpc_request_body(method, json_rpc_port, hyper_json_response, hash, block_blob).await
+            },
+            _ => Err(anyhow!("'{}' not supported", method)),
+        }
+    }
+
+    async fn get_response(json_rpc_port: u16, method: MonerodMethod) -> Result<String, Error> {
+        let full_address = format!("http://127.0.0.1:{}", json_rpc_port);
+        match reqwest::get(format!("{}/{}", full_address, method))
+            .await
+            .unwrap()
+            .json::<serde_json::Value>()
+            .await
+        {
+            Ok(val) => {
+                println!("{}: {}", method, val);
+                Ok(val.to_string())
+            },
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    // The static XMRig request bodies in this test helper function are for testing purposes only. Their content can be
+    // verified for similarity with the actual request bodies that are generated by XMRig as the first log entry in
+    // `applications/minotari_merge_mining_proxy/src/proxy/inner.rs` method `pub(crate) async fn handle(`.
+    async fn json_rpc_request_body(
+        method: MonerodMethod,
+        json_rpc_port: u16,
+        hyper_json_response: bool,
+        hash: Option<String>,
+        block_blob: Option<String>,
+    ) -> Result<String, Error> {
+        let rpc_method = format!("{}", method);
+        let request_body = match method {
+            MonerodMethod::GetBlockTemplate => json!({
+                "jsonrpc": "2.0",
+                "id": "0",
+                "method": rpc_method,
+                "params": {
+                    "wallet_address": DEFAULT_MONERO_ADDRESS,
+                }
+            }),
+            MonerodMethod::GetVersion | MonerodMethod::GetLastBlockHeader => json!({
+                "jsonrpc": "2.0",
+                "id": "0",
+                "method": rpc_method,
+                "params": {}
+            }),
+            MonerodMethod::SubmitBlock => {
+                let block_blob = if let Some(blob) = block_blob {
+                    blob
+                } else {
+                    const ARBITRARY_MERGE_MINING_HASH: &str =
+                        "8e6dab82d22909b40bda27ec0e96aa0c6c0012023d353f3be941eb6ef1793cad";
```

Comment:
what is this used for, this will never pass tari validation?

Reply:
When this code is run as part of an _ignore_ unit test, the merge mining proxy must also be running. To evaluate the proper responses, we insert some merge mining tag here ahead of time as the merge mining proxy will not be able to insert it, as the template is unknown, otherwise the basic validation that checks that one merge mining tag is present will fail.